R/calculateCoex-method.R

```diff
@@ -843,8 +853,284 @@ contingencyTables <- function(objCOTAN, g1, g2) {
+          logLevel = 3L)
+
+  logThis(paste("Calculating", kind, "coex normalization factor"),
+          logLevel = 3L)
+
+  if (isTRUE(actOnCells)) {
+    allNames <- getCells(objCOTAN)
+    normFact <- 1.0 / sqrt(getNumGenes(objCOTAN)) # divided by sqrt(n)
+  } else {
+    allNames <- getGenes(objCOTAN)
+    normFact <- 1.0 / sqrt(getNumCells(objCOTAN)) # divided by sqrt(m)
+  }
+
+  thresholdForPP <- 0.5
+  problematicPairsFraction <- NA
+  if (isTRUE(returnPPFract)) {
+    problematicPairsFraction <-
+      sum(pmin(expectedYY@x, expectedYN@x, expectedNY@x, expectedNN@x) <
+            thresholdForPP) / length(expectedYY@x)
+
+    logThis(paste("Fraction of", kind, "with very low",
+                   "expected contingency tables:",
+                   problematicPairsFraction), logLevel = 3L)
+  }
+
+  coex <- normFact * sqrt(1.0 / pmax(1.0, expectedYY@x) +
+                          1.0 / pmax(1.0, expectedYN@x) +
+                          1.0 / pmax(1.0, expectedNY@x) +
+                          1.0 / pmax(1.0, expectedNN@x))
+
+  rm(expectedYN, expectedNY, expectedNN)
+
+  sqrtTime <- Sys.time()
+  logThis(paste("Calculate", kind, "normalization factor elapsed time:",
+                difftime(sqrtTime, expectedTime, units = "secs")),
+          logLevel = 3L)
+
+  logThis(paste("Retrieving observed", kind, "yes/yes contingency table"),
+          logLevel = 3L)
+
+  c(observedYY, .) %<-%
+    observedContingencyTablesYY(objCOTAN,
+                                actOnCells = actOnCells,
+                                asDspMatrices = TRUE)
+
+  observedTime <- Sys.time()
+  logThis(paste("Observed", kind, "contingency table elapsed time:",
+                 difftime(observedTime, sqrtTime, units = "secs")),
+          logLevel = 3L)
+
+  # coex estimation
+  logThis(paste("Estimating", kind, "coex"), logLevel = 3L)
+
+  coex <- coex * (observedYY@x - expectedYY@x)
+
+  assert_that(2L * length(coex) == length(allNames) * (length(allNames) + 1L),
+              msg = "Output coex@x has the wrong size")
+
+  coex <- new("dspMatrix", Dim = dim(expectedYY), x = coex)
+  rownames(coex) <- colnames(coex) <- allNames
+
+  gc()
+
+  endTime <- Sys.time()
+  logThis(paste("Calculate", kind, "coex elapsed time:",
+                 difftime(endTime, observedTime, units = "secs")),
+          logLevel = 3L)
+
+  logThis(paste("Total calculations elapsed time:",
+                difftime(endTime, startTime, units = "secs")),
+          logLevel = 2L)
+
+  logThis(paste("Calculate", kind, "coex (legacy): DONE"), logLevel = 1L)
+
+  return(list("coex" = coex, "ppf" = problematicPairsFraction))
+}
+
+
+# torch based code
+calculateCoex_Torch <- function(objCOTAN, returnPPFract, deviceStr) {
+  logThis(paste0("Calculate genes coex (torch) on device ", deviceStr,
+                 ": START"), logLevel = 1L)
+
+  startTime <- Sys.time()
+
+  gc()
+  cuda_empty_cache()
+
+  # TODO: 16 bits are OK here?
+  if(deviceStr == "cpu"){
```

Comment:
In the CPU case I would use torch_float() for both, so that the library can choose the right default.